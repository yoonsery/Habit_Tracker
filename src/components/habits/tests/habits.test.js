import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import Habits from '../habits';

describe('Habits component', () => {
  const habits = [
    { name: 'Reading', count: 3, id: 1 },
    { name: 'Eating', count: 0, id: 2 },
  ];

  let HabitsComponent;
  let onIncrement;
  let onDecrement;
  let onDelete;
  let onAdd;
  let onReset;

  beforeEach(() => {
    onIncrement = jest.fn();
    onDecrement = jest.fn();
    onDelete = jest.fn();
    onAdd = jest.fn();
    onReset = jest.fn();

    HabitsComponent = (
      <Habits
        habits={habits}
        onIncrement={onIncrement}
        onDecrement={onDecrement}
        onDelete={onDelete}
        onAdd={onAdd}
        onReset={onReset}
      />
    );
  });

  it('renders', () => {
    const component = renderer.create(HabitsComponent);
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe('Button Click', () => {
    beforeEach(() => {
      render(HabitsComponent);
    });

    it('calls onAdd when clicking "Add" button', () => {
      const input = screen.getByPlaceholderText('new habit');
      const button = screen.getByText('Add');
      const newHabit = 'New Habit';

      userEvent.type(input, newHabit);
      userEvent.click(button);

      expect(onAdd).toHaveBeenCalledWith(newHabit);
    });

    it('calls onIncrement when clicking "increase" button', () => {
      const button = screen.getAllByTitle('increase')[0];
      userEvent.click(button);

      expect(onIncrement).toHaveBeenCalledWith(habits[0]);
    });

    it('calls onDecrement when clicking "decrement" button', () => {
      const button = screen.getAllByTitle('decrease')[0];
      userEvent.click(button);

      expect(onDecrement).toHaveBeenCalledWith(habits[0]);
    });

    it('calls onDelete when clicking "delete" button', () => {
      const button = screen.getAllByTitle('delete')[0];
      userEvent.click(button);

      expect(onDelete).toHaveBeenCalledWith(habits[0]);
    });

    it('calls onReset when clicking "reset" button', () => {
      const button = screen.getByText('Reset All');
      userEvent.click(button);

      expect(onReset).toHaveBeenCalledTimes(1);
    });
  });
});
