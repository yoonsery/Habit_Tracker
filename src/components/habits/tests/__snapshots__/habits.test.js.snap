// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Habits component renders 1`] = `
Array [
  <form
    className="add-form"
    onSubmit={[Function]}
  >
    <input
      className="add-input"
      placeholder="new habit"
      type="text"
    />
    <button
      className="add-button"
    >
      Add
    </button>
  </form>,
  <ul>
    <li
      className="habit"
    >
      <span
        className="habit-name"
      >
        Reading
      </span>
      <span
        className="habit-count"
      >
        3
      </span>
      <div
        className="button-container"
      >
        <button
          className="habit-button habit-increase"
          onClick={[Function]}
          title="increase"
        >
          <i
            className="fas fa-plus-square"
          />
        </button>
        <button
          className="habit-button habit-decrease"
          onClick={[Function]}
          title="decrease"
        >
          <i
            className="fas fa-minus-square"
          />
        </button>
        <button
          className="habit-button habit-delete"
          onClick={[Function]}
          title="delete"
        >
          <i
            className="fas fa-trash"
          />
        </button>
      </div>
    </li>
    <li
      className="habit"
    >
      <span
        className="habit-name"
      >
        Eating
      </span>
      <span
        className="habit-count"
      >
        0
      </span>
      <div
        className="button-container"
      >
        <button
          className="habit-button habit-increase"
          onClick={[Function]}
          title="increase"
        >
          <i
            className="fas fa-plus-square"
          />
        </button>
        <button
          className="habit-button habit-decrease"
          onClick={[Function]}
          title="decrease"
        >
          <i
            className="fas fa-minus-square"
          />
        </button>
        <button
          className="habit-button habit-delete"
          onClick={[Function]}
          title="delete"
        >
          <i
            className="fas fa-trash"
          />
        </button>
      </div>
    </li>
  </ul>,
  <button
    className="habits-reset"
    onClick={[MockFunction]}
  >
    Reset All
  </button>,
]
`;
